### PR TITLE
[NAT, DiNAT] Add backbone class

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the `NatBackbone` and `DinatBackbone` classes, to be used for #20577.

### CODE REVIEW

File: src/transformers/models/dinat/modeling_dinat.py [hunk 1]
```diff
@@ -871,3 +873,122 @@ def forward(
...
+        super().__init__(config)
+
+        requires_backends(self, ["natten"])
+
+        self.stage_names = config.stage_names
+
+        self.embeddings = DinatEmbeddings(config)
+        self.encoder = DinatEncoder(config, self.embeddings.patch_grid)
+
+        self.out_features = config.out_features
```
Comment (sgugger) [inline src/transformers/models/dinat/modeling_dinat.py:893]: The dinat config does not have `out_features` or `stage_names`.

**Review by sgugger — COMMENTED** [review]
Thanks for adding those! Please make sure to merge this after #20675 and #20660, and properly include the changes in those PRs to the new backbones here.

File: src/transformers/models/dinat/modeling_dinat.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/dinat/modeling_dinat.py]: Fixed now, thanks!

File: src/transformers/models/dinat/modeling_dinat.py [hunk 2]
```diff
@@ -21,11 +21,12 @@
 
 import torch
 import torch.utils.checkpoint
-from torch import nn
+from torch import Tensor, nn
```
Comment (sgugger) [inline src/transformers/models/dinat/modeling_dinat.py:24]: Let's use `torch.Tensor` everywhere for more consistency (it is used in several parts of this file).

File: src/transformers/models/dinat/modeling_dinat.py [hunk 3]
```diff
@@ -871,3 +877,120 @@ def forward(
...
+
+    @property
+    def channels(self):
+        return [self.out_feature_channels[name] for name in self.out_features]
+
+    @add_start_docstrings_to_model_forward(DINAT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=BackboneOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values: Tensor,
```
Comment (sgugger) [inline src/transformers/models/dinat/modeling_dinat.py:925]:
```suggestion
        pixel_values: torch.Tensor,
```

File: src/transformers/models/nat/modeling_nat.py [hunk 4]
```diff
@@ -21,11 +21,12 @@
 
 import torch
 import torch.utils.checkpoint
-from torch import nn
+from torch import Tensor, nn
```
Comment (sgugger) [inline src/transformers/models/nat/modeling_nat.py:24]: Same as above.

File: src/transformers/models/nat/modeling_nat.py [hunk 5]
```diff
@@ -849,3 +855,121 @@ def forward(
...
+
+    @property
+    def channels(self):
+        return [self.out_feature_channels[name] for name in self.out_features]
+
+    @add_start_docstrings_to_model_forward(NAT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=BackboneOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values: Tensor,
```
Comment (sgugger) [inline src/transformers/models/nat/modeling_nat.py:903]:
```suggestion
        pixel_values: torch.Tensor,
```

**Review by sgugger — APPROVED** [review]
Good to go with a last couple of nits!

**Pull request merged**